src/sources/util/unix.rs

```diff
@@ -19,3 +33,171 @@ pub fn change_socket_permissions(path: &Path, perms: Option<u32>) -> crate::Resu
+    /// of the process that opened the socket
+    pub peer_creds: bool,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct SocketInode {
+    pub dev: u64,
+    pub ino: u64,
+}
+
+impl Into<Value> for SocketInode {
+    fn into(self) -> Value {
+        let mut map = BTreeMap::new();
+        // really want "as i64" here - actually despite claiming to return
+        // a u64, the value we get from fstat(2) for dev on some platforms
+        // is the bit pattern for -1. So, reinterpret the "u64" as signed.
+        map.insert("dev".to_string(), Value::Integer(self.dev as i64));
+        map.insert("ino".to_string(), Value::Integer(self.ino as i64));
+        Value::Object(map)
+    }
+}
+
+// This is more or less a copy of "struct UCred", but we need to
+// implement Into<Value> for it, and we can't do that to struct UCred itself.
+// in this crate.
+#[derive(Copy, Clone, Debug)]
+pub struct SocketCredentials {
+    // These uid_t/gid_t/pid_t types are from tokio::net::unix. Really, we should
+    // use the libc ones, but the libc crate is not actually declared as a dependency
+    // in Cargo.toml on all platforms.
+    pub uid: uid_t,
+    pub gid: gid_t,
+    pub pid: Option<pid_t>
+}
+
+impl Into<Value> for SocketCredentials {
+    fn into(self) -> Value {
+        let mut map = BTreeMap::new();
+        map.insert("uid".to_string(), Value::Integer(self.uid.into()));
+        map.insert("gid".to_string(), Value::Integer(self.gid.into()));
+        let pid_val = match self.pid {
+            Some(pid) => Value::Integer(pid.into()),
+            None => Value::Null,
+        };
+        map.insert("pid".to_string(), pid_val);
+        Value::Object(map)
+    }
+}
+
+impl From<UCred> for SocketCredentials {
+    fn from(other: UCred) -> Self {
+        Self {
+            uid: other.uid(),
+            gid: other.gid(),
+            pid: other.pid().clone(),
+        }
+    }
+}
+
+/// This structure defines the various kinds of metadata we can
+/// collect off a connected unix-domain socket and expose as source fields.
+#[derive(Clone, Debug)]
+pub struct UnixSocketMetadata {
+    /// The peer address of the socket, as returned from getpeername(2). This
+    /// will usually not be set (unless the connecting peer has explicitly
+    /// bound their socket to a path).
+    pub peer_path: Option<String>,
+
+    /// The inode/device of the socket, as collected from fstat(2). This only
+    /// makes sense for stream sockets, not datagram ones.
+    pub socket_inode: Option<SocketInode>,
+
+    /// The peer credentials of the process which connected to the socket,
+    /// reported by getsockopt(2) SO_PEERCRED/LOCAL_PEERCRED/etc. This may not
+    /// be the same as the process currently writing to the socket, because
+    /// the socket could have been passed down to a child process or sent
+    /// to a different process.
+    pub peer_creds: Option<SocketCredentials>,
+}
```

Comment:
Taking @tobz 's comment (https://github.com/vectordotdev/vector/pull/17705#pullrequestreview-1504103978) into consideration, if the credentials part is removed, we probably don't need to encapsulate the peer_path and inode in this struct, which would also simplify other parts of these changes.